lib/services/webdav.js

```diff
@@ -0,0 +1,613 @@
+    request.session.webdavClientId = null;
+    if (opt_cbk) opt_cbk({success:true, message:"Now logged out."});
+  }
+  else{
+    if (opt_cbk) opt_cbk({success:true, message:"Was not logged in."});
+  }
+}
+
+
+/**
+ * reconnect to the service, or just call cbk() if already connected
+ * @param   request    the connect request
+ * @param   cbk        callback to be called on error (with a err param) or success (no param)
+ * @result  {"success": true}
+ */
+exports.reconnect = function (request, cbk) {
+  var onError = function (opt_err) {
+    opt_err = opt_err || '';
+    console.error('could not reconnect', request.session.webdav_user, opt_err);
+    cbk(opt_err);
+  }
+  if (exports.isLoggedIn(request)){
+    cbk();
+  }
+  else if (request.session.webdav_user){
+    exports.webdavLogin(request, request.session.webdav_user.password, request.session.webdav_user.username, request.session.webdav_user.host, request.session.webdav_user.port,
+        function(){
+          cbk();
+        },
+        onError);
+  }
+  else{
+    onError();
+  }
+}
+
+
+/**
+ * answer that the user is not connected
+ * @param   cbk        the middle ware's callback
+ */
+exports.errNotLoggedIn = function (cbk) {
+  cbk({
+    success:false,
+    message:"User not connected yet. You need to call the 'login' service first.",
+    code: 401
+  });
+}
+
+/**
+ * Load the data associated with the current user account
+ * Call the provided callback with this data
+ *		status		: {"success": true},
+ *		data 		{
+ * 						display_name: "Alexandre Hoyau",
+ * 						quota_info: {
+ * 						available: 5368709120,
+ * 						used: 144201723
+ * 					}
+ */
+exports.getAccountInfo = function (request, response, next, cbk) {
+  if(!request || !request.session || !request.session.webdav_user)
+  {
+    exports.errNotLoggedIn(cbk);
+  }
+  else {
+    cbk({"success": true},
+        {
+          display_name: request.session.webdav_user.username
+        });
+  }
+}
+
+
+// ******* commands
+
+/**
+ * List the files of a given folder
+ * @result 	an object like this one:
+ * {
+ *   "status": {
+ *     "success": true
+ *   },
+ *   "data": [
+ *     {
+ *       "bytes": 0,
+ *       "modified": "Thu, 03 Jan 2013 14:24:53 +0000",
+ *       "title": "name",
+ *       "is_dir": true,
+ *     },
+ *
+ *     ...
+ *   ]
+ * }
+ *
+ */
+exports.ls = function (path, request, response, next, cbk) {
+  // not connected already
+  if(!webdavClientsArray[request.session.webdavClientId])
+  {
+    return cbk({success:false, message: 'not connected'});
+  }
+
+  var dav = webdavClientsArray[request.session.webdavClientId];
+
+  dav.list(path, function(err, data) {
+
+    path = path.replace(/\/+/g,'/'); // Remove doble / on the path
+    path = path.replace(/\/$/,'')+'/'; // Force to have last / on the path
+    if (err!==false){
+      filesData=[];
+      parseString(data, function (err, result) {
+        if(result['d:multistatus'] && result['d:multistatus']['d:response'])
```

Comment:
Handle the case where `err` is not null here
`if(!err && result['d:multistatus'] && result['d:multistatus']['d:response'])`